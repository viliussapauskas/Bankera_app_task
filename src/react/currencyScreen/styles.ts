import { StyleSheet } from 'react-native'

export const styles = StyleSheet.create({
    title: {
        textAlign: 'center',
    },
    returnValue: {
        alignItems: 'center',
        marginTop: 40,
    },
})
