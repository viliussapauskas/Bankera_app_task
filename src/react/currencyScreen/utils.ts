export const currencyToUSD = (
    ammountOfCurrency: string,
    cryptoValue: number
): number | null => {
    const parsedAmmountOfCurrency = parseFloat(ammountOfCurrency)
    if (isNaN(parsedAmmountOfCurrency)) {
        return null
    }
    const response = parsedAmmountOfCurrency / cryptoValue
    return Math.round(response * 100) / 100
}
