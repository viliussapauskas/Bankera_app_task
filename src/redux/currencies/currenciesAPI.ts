import axios from 'axios'
import { GET_CURRENCIES_URLS } from '../../config'
import { Currency } from './types'

export async function fetchCurrencies(): Promise<Currency[]> {
    var axiosCalls = GET_CURRENCIES_URLS.map((url) => axios.get(url))

    return axios.all(axiosCalls).then(
        axios.spread((...allData) => {
            let currenciesResponse: Currency[] = []

            allData.map((response) => {
                currenciesResponse.push({
                    id: response.data.currencyFrom,
                    name: response.data.currencyFrom,
                    value: response.data.last,
                })
            })
            return currenciesResponse
        })
    )
}
