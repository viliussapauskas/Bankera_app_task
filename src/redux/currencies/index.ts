export { fetchCurrencies } from './currenciesAPI'
export {
    currenciesReducer,
    currenciesActionCreators,
    currenciesSlice,
    getCurrenciesStateSelector,
    loadCurrenciesAsync,
} from './currenciesSlice'
export { initialState } from './state'
export { type Currency } from './types'
