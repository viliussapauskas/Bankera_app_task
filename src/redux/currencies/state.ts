import { CurrenciesState } from './types'

export const initialState: CurrenciesState = {
    value: undefined,
    status: 'idle',
}
