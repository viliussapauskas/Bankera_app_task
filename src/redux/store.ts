import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit'
import { currenciesReducer } from './currencies'

export const store = configureStore({
    reducer: {
        currencies: currenciesReducer,
    },
})

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    Action<string>
>
