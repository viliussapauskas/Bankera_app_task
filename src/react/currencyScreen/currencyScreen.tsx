import React, { FC, useState } from 'react'
import { View, Text } from 'react-native'
import { TextInput, Title } from 'react-native-paper'
import { getCurrencyByNameSelector } from '../../redux/currencies'
import { useAppSelector } from '../../redux/hooks'
import { Layout } from '../components'

export interface CurrencyScreenProps {
    route: {
        params: {
            name: string
        }
    }
}

export const CurrencyScreen: FC<CurrencyScreenProps> = ({ route }) => {
    const currency = useAppSelector(
        getCurrencyByNameSelector(route?.params?.name)
    )

    const [ammountOfCurreny, setAmmountOfCurreny] = useState()

    return (
        <Layout>
            <View>
                <TextInput
                    label={currency?.name}
                    value={ammountOfCurreny}
                    // type="number"
                    numberOfLines={1}
                    keyboardType="number-pad"
                    onChangeText={(text) => setAmmountOfCurreny(text)}
                />
                <Text>
                    {route?.params?.name} --- {currency?.value}
                    {/* Your return: */}
                </Text>
                <Title>
                    Your return:{' '}
                    {ammountOfCurreny
                        ? ammountOfCurreny * currency?.value!
                        : undefined}
                </Title>
            </View>
        </Layout>
    )
}
