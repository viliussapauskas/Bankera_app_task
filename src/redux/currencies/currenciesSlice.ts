import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { RootState } from '../store'
import { fetchCurrencies } from './currenciesAPI'
import { initialState } from './state'
import { Currency } from './types'

export const loadCurrenciesAsync = createAsyncThunk<Currency[]>(
    'currencies/fetchCurrencies',
    async () => await fetchCurrencies()
)

export const currenciesSlice = createSlice({
    name: 'currencies',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder
            .addCase(loadCurrenciesAsync.pending, (state) => {
                state.status = state.value !== undefined ? 'idle' : 'loading'
            })
            .addCase(loadCurrenciesAsync.fulfilled, (state, action) => {
                state.status = 'idle'
                state.value = action.payload
            })
            .addCase(loadCurrenciesAsync.rejected, (state) => {
                state.status = 'failed'
            })
    },
})

const { reducer, actions } = currenciesSlice
export const currenciesReducer = reducer
export const currenciesActionCreators = actions

export const getCurrenciesStateSelector = (state: RootState) => state.currencies
export const getCurrencyByNameSelector = (name: string) => (state: RootState) =>
    state.currencies.value?.find((x) => x.name === name) ?? undefined
