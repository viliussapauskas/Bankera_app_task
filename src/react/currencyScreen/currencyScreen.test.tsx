import { CurrencyScreen } from './currencyScreen'
import { fireEvent, render } from '@testing-library/react-native'
import { Provider } from 'react-redux'
import { Currency } from '../../redux/currencies'
import * as hooks from '../../redux/hooks'
import { store } from '../../redux/store'

describe('WalletScreen component tests', () => {
    type Props = React.ComponentProps<typeof CurrencyScreen>

    const getDefaultProps = (): Props => {
        return {
            route: {
                params: {
                    name: 'BitCoin',
                },
            },
        }
    }

    const getComponent = (props: Props = getDefaultProps()) => {
        return render(
            <Provider store={{ ...store }}>
                <CurrencyScreen {...props} />
            </Provider>
        )
    }

    beforeEach(() => {
        jest.resetAllMocks()
    })
    beforeAll(() => {
        jest.useFakeTimers()
    })

    it('Should render CurrencyScreen component', () => {
        const component = getComponent()
        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Should render correct title', () => {
        jest.spyOn(hooks, 'useAppSelector').mockImplementation(() => {
            return {
                id: 'BTC',
                name: 'BitCoin title input',
                value: 0.5,
            } as Currency
        })
        const component = getComponent()
        expect(component.toJSON()).toMatchSnapshot()
    })

    // it('Should not render return card when returnValue empty', () => {
    //     const { getAllByTestId } = getComponent()
    //     expect(getAllByTestId('currencyScreen.returnCard').length).toBe(0)
    // })

    it('Should render return card when input value is correctly', async () => {
        const { getByTestId, debug, findByTestId } = getComponent()
        fireEvent.changeText(getByTestId('currencyScreen.ammountInput'), '1.2')
        console.log(findByTestId('currencyScreen.returnCard'))

        const returnCard = await findByTestId('currencyScreen.returnCard')

        console.log(returnCard, 'RETURN CARD')
        // expect(return)
        // debug()
    })
})
