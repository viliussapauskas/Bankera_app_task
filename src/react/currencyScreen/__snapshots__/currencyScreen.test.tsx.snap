// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyScreen component tests Should render CurrencyScreen component 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
  testID="layout"
>
  <RCTScrollView
    accessible={false}
    contentContainerStyle={
      Object {
        "flex": 1,
        "padding": 20,
      }
    }
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
  >
    <RCTRefreshControl />
    <View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
        testID="currencyScreen"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "textAlign": "left",
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "textAlign": "center",
                  },
                ],
              ],
            ]
          }
          testID="currencyScreen.title"
        >
          BitCoin title input
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "textAlign": "left",
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.54)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 12,
                    "letterSpacing": 0.4,
                    "lineHeight": 20,
                    "marginVertical": 2,
                  },
                  undefined,
                ],
              ],
            ]
          }
        >
          * Enter ammount of selected currency to see your return in USD
        </Text>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgb(231, 231, 231)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              Object {},
            ]
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.26)",
                "bottom": 0,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scaleY": 0.5,
                  },
                ],
                "zIndex": 1,
              }
            }
          />
          <View
            style={
              Array [
                Object {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                Object {
                  "minHeight": 64,
                },
              ]
            }
          >
            <View
              collapsable={false}
              pointerEvents="none"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                  ],
                  "zIndex": 3,
                }
              }
            >
              <Text
                collapsable={false}
                maxFontSizeMultiplier={1.5}
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "#6200ee",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 34,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                BitCoin title input
              </Text>
              <Text
                collapsable={false}
                maxFontSizeMultiplier={1.5}
                numberOfLines={1}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 34,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                BitCoin title input
              </Text>
            </View>
            <TextInput
              editable={true}
              keyboardType="numeric"
              maxFontSizeMultiplier={1.5}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder=""
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              selectionColor="#6200ee"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                  },
                  Object {
                    "paddingLeft": 12,
                    "paddingRight": 12,
                  },
                  Object {
                    "height": 64,
                  },
                  Object {
                    "paddingBottom": 4,
                    "paddingTop": 24,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  Array [
                    Object {},
                  ],
                ]
              }
              testID="currencyScreen.ammountInput"
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
